sdks/go/pkg/beam/io/spannerio/test_utils.go

```diff
@@ -0,0 +1,67 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spannerio
+
+import (
+	"cloud.google.com/go/spanner"
+	db "cloud.google.com/go/spanner/admin/database/apiv1"
+	"cloud.google.com/go/spanner/spannertest"
+	"context"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc"
+	"testing"
+)
+
+type TestDto struct {
+	One string `spanner:"One"`
+	Two int64  `spanner:"Two"`
+}
+
+func newServer(t *testing.T) (*spannertest.Server, func()) {
+	srv, err := spannertest.NewServer("localhost:0")
+	if err != nil {
+		t.Fatalf("Starting in-memory fake spanner: %v", err)
+	}
+
+	return srv, func() {
+		srv.Close()
+	}
+}
+
+func createFakeClient(address string, database string) (*spanner.Client, *db.DatabaseAdminClient, func(), error) {
```

Comment:
We can simplify things for the testing since Spanner has an env variable override we can use.

See `SPANNER_EMULATOR_HOST` in the spannertest docs.
https://pkg.go.dev/cloud.google.com/go/spanner/spannertest#section-readme

This lets us use the normal client paths, and may be able to work better on portable runners at least for unit test purposes.

Alternatively, we add another exported field to the spannerFn to put in the server address, and when that's present, we provide the appropriate connection option for the fake override. While this does move some test specific code into the production code, it makes testing much easier.

Reply:
I've stuck with this approach for admin clients (due to no support for the env var) and in unit tests to support parallelism (is that really a concern?).

For the integration test preferred SPANNER_EMULATOR_HOST.

Are we ok with this approach?